File: 8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python
#
# [8] 字符串转换整数 (atoi)
#
# 解法1(T90% S65%)：还是按照题意一步一步写就好，首先通过lstrip()去除前面的空格，然后判断下数字的正负号，依次遍历字符串直到遇到非数字，把所有数字字符都保存到一个数组中，如果数组为空则直接返回0，否则将数组拼接字符串并转为整数，最后再根据int范围截断一下

# @lc code=start
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        s = s.lstrip()
        if len(s) > 0:
            if s[0] in ['+', '-']:
                flag = s[0]
                s = s[1:]
            else: flag = '+'

        res = []
        for c in s:
            if c.isnumeric(): res.append(c)
            else: break

        if len(res) == 0: return 0

        num = int("".join(res))
        if flag == '+':
            if num > 2**31-1: return 2**31-1
        else:
            num = -num
            if num < -2**31: return -2**31
        return num
# @lc code=end

File: javascript/8.字符串转换整数-atoi.js
/*
 * @lc app=leetcode.cn id=8 lang=javascript
 *
 * [8] 字符串转换整数 (atoi)
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function(s) {
    if (s.length == 0) { return 0; }

    s = s.trimStart();
    switch (s[0]) {
        case "-":
            var flag = true;
            s = s.slice(1);
            break;
        case "+":
            var flag = false;
            s = s.slice(1);
            break;
        case "0": case "1": case "2":case "3": case "4": case "5":case "6": case "7": case "8": case "9":
            var flag = false;
            break;
        default:
            return 0;
    }

    var i=0;
    while (i<s.length && (s[i]>="0" && s[i]<="9")) { i++; }
    if (i==0) { return 0; }     // 如果第一个有用字符就非法

    var res = parseInt(s.slice(0, i));
    if (flag) { res = -res; }
    if (res < -Math.pow(2, 31)) { return -Math.pow(2, 31); }
    else if (res > Math.pow(2, 31)-1) { return Math.pow(2, 31)-1; }
    return res;
};
// @lc code=end

